File: exercises/capstone/ex03_start_mode/ex03_start_mode.cpp
// Exercise 03 -- localization and start mode (Step 3).
//
// Read exercises/capstone/ex03_start_mode/README.md first.
#include "planning/airport/taxi_planner.hpp"

namespace planning::airport {

const char* toString(StartMode mode) {
  // TODO(you): one word per mode, for the demo and the failure messages.
  (void)mode;
  return "?";
}

Localization localize(const ZoneLayer& layer, const TaxiGraph& gated, const AircraftModel& aircraft,
                      const Pose& pose, const Localization* previous) {
  // TODO(you):
  //
  // 1. classifyFootprint().  Refuse (StartMode::Fault) when a wingtip is
  //    inside a structure, when any tyre is off load-bearing pavement, or when
  //    the three tyres disagree about which zone they are in.  Put the reason
  //    in `detail` -- a refusal nobody can read is not much of a refusal.
  //
  // 2. Map the gear zone to a mode:  stand -> Stand, apron and de-icing ->
  //    Apron, taxiway -> TaxiwayCapture, runway *and runway protected* ->
  //    Runway.  A protected area is treated as runway: no free-space planning.
  //
  // 3. Hysteresis.  When `previous` is supplied and the mode has changed, keep
  //    the old mode until the footprint is kModeHysteresis metres inside the
  //    new zone.  Otherwise a metre of position noise flips the mode between
  //    apron and taxiway on successive replans.
  //
  // 4. Find the nearest guidance line whose zone matches, and fill in
  //    nearestEdge, nearestS, crossTrack, headingError and onGuidanceLine.
  //    In Stand mode pick the direction that leads *out* of the stand, not the
  //    one closest to the current heading: that is what makes a nose-in
  //    parking position show up as a heading error of 180 degrees, which is
  //    how the pipeline knows to ask for a pushback.
  (void)layer;
  (void)gated;
  (void)aircraft;
  (void)pose;
  (void)previous;
  return Localization{};
}

}  // namespace planning::airport

File: src/capstone/problems/ex03-start-mode.ts
// Exercise 03 -- localization and start mode (Step 3).
// Brief: docs/capstone/ex03-start-mode.md

import {
  type AircraftModel,
  centroidOf,
  crossTrack,
  distanceToBoundary,
  type Pose,
  reversed,
  type TaxiGraph,
  wrapAngle,
  type ZoneLayer,
} from '../../airport/index.js';
import { emptyLocalization, type Localization, MODE_HYSTERESIS } from '../types.js';
import { classifyFootprint } from './ex01-zone-layer.js';

export function localize(
  layer: ZoneLayer,
  gated: TaxiGraph,
  aircraft: AircraftModel,
  pose: Pose,
  previous?: Localization,
): Localization {
  // TODO(you): where is the aeroplane, and what kind of planning does that
  // allow?  In order:
  //
  //   1  classifyFootprint().  A wingtip in a structure, a tyre off
  //      load-bearing pavement, or three tyres that disagree about the zone are
  //      each a Fault -- and each needs a `detail` a human can act on.
  //   2  The zone the three tyres agree on decides the mode.  A protected area
  //      counts as runway: no free-space planning between the holding position
  //      and the runway edge.
  //   3  Hysteresis.  If `previous` is given and the mode would change, only
  //      let it change once the whole footprint is MODE_HYSTERESIS metres
  //      inside the new zone.  Without this, a metre of position noise flips
  //      the planner between apron and taxiway on successive replans.
  //   4  The nearest guidance line *of this zone*, and which way along it.  In
  //      stand mode that is always the way that leads OUT of the stand -- which
  //      is what makes a nose-in parking position show up as a 180 degree
  //      heading error rather than as a separate flag.
  //   5  crossTrack, headingError, and whether that is inside the capture
  //      window (2 m and 10 degrees).
  void layer;
  void gated;
  void aircraft;
  void previous;
  void classifyFootprint;
  void MODE_HYSTERESIS;
  void centroidOf;
  void crossTrack;
  void distanceToBoundary;
  void reversed;
  void wrapAngle;
  return emptyLocalization(pose);
}
